kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/util/ServiceLocator.java

```diff
@@ -89,20 +95,92 @@ public static ServiceLocator getInstance() {
      *            the service to locate
      * @param function
      *            the function to execute
+     * @throws GwtKuraException
+     *             if the service function throws an exception
      * @return the return value of the function
      */
-    public static <T, R> R withOptionalService(final Class<T> serviceClass, final ServiceFunction<T, R> function) {
+    public static <T, R> R withOptionalService(final Class<T> serviceClass, final ServiceFunction<T, R> function)
+            throws GwtKuraException {
         final BundleContext ctx = FrameworkUtil.getBundle(ServiceLocator.class).getBundleContext();
         final ServiceReference<T> ref = ctx.getServiceReference(serviceClass);
-        if (ref == null) {
-            return function.apply(null);
+
+        try {
+            if (ref == null) {
+                return function.apply(null);
+            }
+
+            final T service = ctx.getService(ref);
+            try {
+                return function.apply(service);
+            } finally {
+                ctx.ungetService(ref);
+            }
+        } catch (Exception e) {
+            throw new GwtKuraException(GwtKuraErrorCode.INTERNAL_ERROR, e);
         }
+    }
+
+    /**
+     * Lookup services and iterate over their instances
+     *
+     * @param serviceClass
+     *            the services to look for
+     * @param consumer
+     *            the consumer which will be called for each service instance
+     * @throws GwtKuraException
+     *             if any service consumer throws an exception
+     */
+    public static <T> void withAllServices(final Class<T> serviceClass, final ServiceConsumer<T> consumer)
```

Comment:
well, withOptionalService was added by commit 8a5c94415035b8b5256fd2969eda38a2e019f383 and, for what I know, is not released. So, we are eventually still in time to rename it 😉